sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java

```diff
@@ -470,23 +469,50 @@ private static Exception instantiateUnexpectedException(final UnexpectedExceptio
+        final Class<? extends Response<?>> cls = (Class<? extends Response<?>>) TypeUtil.getRawClass(entityType);
+
+        final HttpResponse httpResponse = response.getSourceResponse();
+        final HttpRequest request = httpResponse.getRequest();
+        final int statusCode = httpResponse.getStatusCode();
+        final HttpHeaders headers = httpResponse.getHeaders();
+        final Object decodedHeaders = response.getDecodedHeaders();
+        // Inspection of the response type needs to be performed to determine which course of action should be taken to
+        // instantiate the Response<?> from the HttpResponse.
+        //
+        // If the type is either the Response or PagedResponse interface from azure-core a new instance of either
+        // ResponseBase or PagedResponseBase can be returned.
         if (cls.equals(Response.class)) {
-            cls = (Class<? extends Response<?>>) (Object) ResponseBase.class;
+            // For Response return a new instance of ResponseBase cast to the class.
+            return Mono.defer(() -> Mono.just(cls.cast(new ResponseBase<>(request, statusCode, headers, bodyAsObject,
+                decodedHeaders))));
         } else if (cls.equals(PagedResponse.class)) {
-            cls = (Class<? extends Response<?>>) (Object) PagedResponseBase.class;
-
-            if (bodyAsObject != null && !TypeUtil.isTypeOrSubTypeOf(bodyAsObject.getClass(), Page.class)) {
-                return monoError(logger, new RuntimeException(MUST_IMPLEMENT_PAGE_ERROR));
-            }
+            // For PagedResponse return a new instance of PagedResponseBase cast to the class.
+            //
+            // PagedResponse needs an additional check that the bodyAsObject implements Page.
+            //
+            // If the bodyAsObject is null use the constructor that take items and continuation token with null.
+            // Otherwise, use the constructor that take Page.
+            return Mono.create(sink -> {
+                if (bodyAsObject != null && !TypeUtil.isTypeOrSubTypeOf(bodyAsObject.getClass(), Page.class)) {
+                    sink.error(logger.logExceptionAsError(new RuntimeException(MUST_IMPLEMENT_PAGE_ERROR)));
+                } else if (bodyAsObject == null) {
+                    sink.success(cls.cast(new PagedResponseBase<>(request, statusCode, headers, null, null,
+                        decodedHeaders)));
+                } else {
+                    sink.success(cls.cast(new PagedResponseBase<>(request, statusCode, headers, (Page<?>) bodyAsObject,
+                        decodedHeaders)));
+                }
+            });
         }
 
-        final Class<? extends Response<?>> clsFinal = cls;
-        return Mono.just(RESPONSE_CONSTRUCTORS_CACHE.get(clsFinal))
+        // Otherwise, rely on reflection, for now, to get the best constructor to use to create the Response sub-type.
```

Comment:
Curious, in what scenarios is the return type not Response<T> or PageResponse<T>. For instance in ACR I never saw a response different so I am guessing this is a scenario other than track2 client libraries.

Reply:
This would an example of a service API that returns a non-Response, non-PagedResponse type: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/implementation/FormRecognizerClientImpl.java#L175

Reply:
[Discussed it offline] We create strongly typed response objects for scenarios that have strongly typed headers.